metricbeat/scripts/mage/docs_collector.go

```diff
@@ -0,0 +1,508 @@
+
+}
+
+// gatherMetricsets gathers all the metricsets for a given module
+func gatherMetricsets(modulePath string, moduleName string) ([]metricsetData, error) {
+	metricsetList, err := filepath.Glob(filepath.Join(modulePath, "/*"))
+	if err != nil {
+		return nil, err
+	}
+	var metricsets []metricsetData
+	for _, metricset := range metricsetList {
+		isMetricset, err := testIfDocsInDir(metricset)
+		if err != nil {
+			return nil, err
+		}
+		if !isMetricset {
+			continue
+		}
+		metricsetName := filepath.Base(metricset)
+		release, err := getReleaseState(filepath.Join(metricset, "_meta/fields.yml"))
+		if err != nil {
+			return nil, err
+		}
+
+		// generate the asciidoc link used in the module docs, since we need this in a few places
+		link := fmt.Sprintf("<<metricbeat-metricset-%s-%s,%s>>", moduleName, metricsetName, metricsetName)
+
+		// test to see if the metricset has a data.json
+		hasData := false
+		_, err = os.Stat(filepath.Join(metricset, "_meta/data.json"))
+		if err == nil {
+			hasData = true
+		}
+
+		ms := metricsetData{
+			Path:       metricset,
+			Title:      metricsetName,
+			Release:    release,
+			Link:       link,
+			DataExists: hasData,
+		}
+
+		metricsets = append(metricsets, ms)
+
+	} // end of metricset loop
+
+	return metricsets, nil
+}
+
+// gatherData gathers all the data we need to construct the docs that end up in metricbeat/docs
+func gatherData(modules []string) (map[string]moduleData, error) {
+	moduleMap := make(map[string]moduleData)
+	//iterate over all the modules, checking to make sure we have an asciidoc file
+	for _, module := range modules {
+
+		isModule, err := testIfDocsInDir(module)
+		if err != nil {
+			return moduleMap, err
+		}
+		if !isModule {
+			continue
+		}
+		moduleName := filepath.Base(module)
+
+		err = createDocsPath(moduleName)
+		if err != nil {
+			return moduleMap, err
+		}
+
+		fieldsm, err := loadModuleFields(filepath.Join(module, "_meta/fields.yml"))
+		if err != nil {
+			return moduleMap, err
+		}
+
+		cfgPath, err := getConfigfile(module)
+		if err != nil {
+			return moduleMap, err
+		}
+
+		metricsets, err := gatherMetricsets(module, moduleName)
+		if err != nil {
+			return moduleMap, err
+		}
+
+		//dump the contents of the module asciidoc
+		moduleDoc, err := ioutil.ReadFile(filepath.Join(module, "_meta/docs.asciidoc"))
+		if err != nil {
+			return moduleMap, err
+		}
+
+		fieldsm.Path = module
+		fieldsm.CfgFile = cfgPath
+		fieldsm.Metricsets = metricsets
+		fieldsm.Asciidoc = string(moduleDoc)
+		fieldsm.Dashboards = hasDashboards(module)
+
+		moduleMap[moduleName] = fieldsm
+
+	} // end of modules loop
+
+	return moduleMap, nil
+}
+
+// writeModuleDocs writes the module-level docs
+func writeModuleDocs(modules map[string]moduleData) error {
+	for moduleName, mod := range modules {
+
+		var moduleFile bytes.Buffer
+		moduleFile.WriteString(generatedNote + fmt.Sprintf("[[metricbeat-module-%s]]\n== %s module\n\n", moduleName, mod.Title))
```

Comment:
I think it'd be more readable (and easier to modify in the future) if these lines are not combined in a single `WriteString`.